R/histogram_to_rgbwt.R

```diff
@@ -20,42 +20,45 @@
 #'
 #' Colorize given histogram with input palette.
 #'
-#' @param fhistogram Matrix or array R datatype interpreted as histogram.
+#' @param fhistogram Matrix or 2D array with the histogram of values.
 #'
-#' @param RGBA matrix (4xn dim, n>= 2) with R, G, B and alpha channels
-#'             in integers, defaults to shades of `red`, `green` and `blue` with `alpha = 255`.
+#' @param RGBA 4-by-N matrix floating-point R, G, B and A channels for the palette. Overrides `col`.
+#'
+#' @param col Colors to use for coloring.
+#'
+#' @param zlim Values to use as extreme values of the histogram
 #'
 #' @return RGBWT matrix.
 #'
 #' @export
 #' @useDynLib scattermore, .registration=TRUE
 
 histogram_to_rgbwt <- function(fhistogram,
-                               RGBA = array(c(250, 128, 114, 255, 144, 238, 144, 255, 176, 224, 230, 255), c(4, 3))) {
-  if (!is.matrix(fhistogram) && !is.array(fhistogram)) stop("fhistogram in matrix form expected")
-  if (dim(fhistogram)[2] < 2) stop("not fhistogram format")
-  if (dim(RGBA)[1] != scattermore.globals$dim_RGBA) stop("RGBA with 4 rows expected")
-  if (dim(RGBA)[2] < 2) stop("RGBA with at least 2 colors expected")
+                               RGBA = col2rgb(col, alpha = T),
+                               col = c("white", "black"),
+                               zlim = c(min(fhistogram), max(fhistogram))) {
+  if (!is.matrix(fhistogram) && !is.array(fhistogram)) stop("unsupported histogram format")
+  if (length(dim(fhistogram)) != 2) stop("unsupported histogram format")
+  if (dim(RGBA)[1] != 4) stop("RGBA with 4 rows expected")
+  if (dim(RGBA)[2] < 2) stop("at least 2-color palette is required")
 
   rows <- dim(fhistogram)[1]
   cols <- dim(fhistogram)[2]
-  size <- dim(RGBA)[2]
+  pal_size <- dim(RGBA)[2]
 
-  RGBWT <- rep(0, rows * cols * scattermore.globals$dim_RGBWT)
+  RGBWT <- array(0, c(rows, cols, 5))
 
-  minimum <- min(fhistogram)
-  maximum <- max(fhistogram)
-  # normalize histogram on values 0-1
-  normalized_fhistogram <- (fhistogram - minimum) / max((maximum - minimum), scattermore.globals$epsilon)
+  normalized_fhistogram <- pmin(pal_size, pmax(
+    0,
+    pal_size * (fhistogram - zlim[1]) / max((zlim[2] - zlim[1]), scattermore.globals$epsilon)
+  ))
 
   result <- .C("histogram_to_rgbwt",
-    dimen = as.integer(c(rows, cols, size)),
+    dimen = as.integer(c(rows, cols, pal_size)),
     fRGBWT = as.single(RGBWT),
     RGBA = as.single(RGBA / 255),
-    normalized_fhistogram = as.single(normalized_fhistogram)
+    normalized_fhistogram = as.integer(normalized_fhistogram)
```

Comment:
should't there be as.single? i am not sure if we should change the type parameter for histogram... the result in one example isn't correct, it should be a colorful image... so i think there could be a mistake?
<img width="471" alt="image" src="https://github.com/exaexa/scattermore/assets/56218654/a6f34805-ac20-4bef-8a16-d372884ba671">

Reply:
I actually vastly simplified the default colors so yeah, that's correct.

By converting to integers directly in R I avoid the size_t UB later.

Reply:
anyway the colors are back (just wanted to avoid specifying custom ones so I'm going with the hcl.colors)

Reply:
(current version) 
![image](https://github.com/exaexa/scattermore/assets/271543/fbb2da5f-9234-4ee4-b719-196ad7ee2c8d)

Reply:
ou, okay, this seems correct, I got confused that the default parameters were different :D